### ISSUE TITLE
are pings handled in async mode?

### ISSUE THREAD

**Issue by horacimacias** [snapview/tokio-tungstenite#88]
Looking at https://github.com/snapview/tokio-tungstenite/issues/6 it looks like pings should be already supported and responded to when using tungstenite as a client. 
Is that correct?

I wrote a ws client and I see my server sending a Ping but never getting a response from the client.
Is there any way I can confirm pings are being replied to?
Is there a "connect and wait forever" sample client I can look at to check what am I doing wrong?

Issue is most likely on my side; if you have any suggestions on how to troubleshoot I'd appreciate.

**Comment by agalakhov** [comment]
Yes, pings are supported. Even in async mode Tungstenite replies to pings automatically and immediately. Ping behavior is covered by standard Autobahn tests. The client you're looking for is `autobahn-client` from Tungstenite examples.
It is safe just to ignore pings as Tungstenite handles them for you.

**Comment by horacimacias** [comment]
thanks!

This is likely a silly question but is it possible that pings are only replied to if my client reads anything from the socket?

My sample client was made of 2 peers, one sender-only (websocket 1) and one receiver-only (websocket 2).
On my sender-only code I was not doing any read from the websocket and I see on the server there are no ping/pongs for that client.
I updated my code to read and send pings when received, as I thought this crate was not replying to pings, and on the server then I saw 2 pongs, likely the one generated internally and the one created by me.
 Does this make sense?

**Comment by sdroege** [comment]
You need to read from the read-side of the socket so that it receives/handles pings, and on the next write it would then send the corresponding pong.

**Comment by kellytk** [comment]
> You need to read from the read-side of the socket so that it receives/handles pings

Is that documented?  As IMHO it contradicts:

> Even in async mode Tungstenite replies to pings automatically and immediately

**Comment by agalakhov** [comment]
This is documented and belongs to Tokio. Tungstenite does nothing unless it is called. There are no things like background threads. If you don't call any of Tunstenite APIs, it does nothing. It  just won't be there.

In WebSocket you could receive any messages at any time. Not only pings. Your application has to either poll WebSocket periodically or start a background job waiting for messages in order to handle them. If you do anything like that, Tungstenite will handle pings for you.

**Comment by horacimacias** [comment]
maybe my following comment is picky and/or incorrect, feel free to ignore.

In my case I was starting with a minimum test application that was opening the websocket and not writing/reading from it. My server was tearing down the connection because of missed pings, which I understood Tungstenite would _automatically and immediately_ handle.
That didn't seem to be the case.

Then I added code to read from websocket and reply to pings from my test application. Then I noticed on the server I had two replies to pings, 1 presumably generated by Tungstenite and the other one by my application explicitly.
I removed the explicit replies and any socket writing in my application and I still saw ping replies.

Long story short, here's my understanding so far. Please let me know if this is incorrect.

* Does Tungstenite handle pings? **If/when you read from the websocket**, yes it will handle then and reply to them.
* Do I need to do **any** websocket write for the ping to be handled/answered? No. as long as you **read** from the websocket, Tungstenite will write the response for you regardless of you ever using the websocket for writing or not.

**Comment by application-developer-DA** [comment]
> Does Tungstenite handle pings? If/when you read from the websocket, yes it will handle then and reply to them.

Yes, it does. But the reading part should make some progress, so that the underlying functions which read from the socket (and write the response) do some progress.

> Do I need to do any websocket write for the ping to be handled/answered? No. as long as you read from the websocket, Tungstenite will write the response for you regardless of you ever using the websocket for writing or not.

Right (but only for this particular case). In case you want to write some other messages explicitly, you have to make sure that the corresponding tasks make some progress. Normally this is ensured in "typical" server applications, i.e. when you split reading and writing part and there are tasks which use them.

**Comment by application-developer-DA** [comment]
@horacimacias it seems like the original question has been answered, so I'll close the issue (it just easier to track what is still open when the closed issues are closed ;), but in case there are any issues, feel free to re-open this one or open a new one.
